### DISCUSSION TITLE
Hydrogen sampler/audio engine as a standalone or plugin

### DISCUSSION THREAD

**Discussion by dingodoppelt** [hydrogen-music/hydrogen#1467] [Ideas]
Hi everybody,
hydrogen is my go-to-sampler for use with a padcontroller for fingerdrumming. I don't ever use the sequencer and would love to be able to just load a drumkit, preferably as an lv2 plugin, and play away without launching the fully featured hydrogen. Since no plugin I tried that claims to load hydrogen kits really did the job (either no round robins, layers or other missing features I can't live without) I figured how hard it would be to isolate some source files and just use the code necessary to load a kit and to play back samples and to wrap that in an lv2 plugin which takes midi input and outputs stereo audio, no GUI needed.
I'm not a programmer but I had a naive look at the code and thought about using just an instance of AudioEngine and to remove any leftover dependencies to GUI and sequencer to be left with just the sampler and drumkit classes.
Would that even be possible or is there maybe some plugin that already does what I envision here?
cheers, nils

**Comment by theGreatWhiteShark** [comment]
> I figured how hard it would be to isolate some source files and just use the code necessary to load a kit and to play back samples and to wrap that in an lv2 plugin which takes midi input and outputs stereo audio, no GUI needed.

Well, that sounds like it would definitely take more than one weekend.

One member of the dev team (@mauser) [started](https://github.com/hydrogen-music/hydrogen/tree/H2asLV2) implementing a LV2 version of Hydrogen. But I do not know how much was done there or why he stopped.

> I'm not a programmer but I had a naive look at the code and thought about using just an instance of AudioEngine and to remove any leftover dependencies to GUI and sequencer to be left with just the sampler and drumkit classes.

Starting with the GUI version if no GUI is needed will probably be a pain. There are two other versions of Hydrogen that might fit your needs a lot better: `h2cli` - a lightweight command line interface - and `h2player` - which plays back a `.h2song` files. Both of them are neither in development anymore nor heavily used. But last time I checked they were at least still working.

Also, you shouldn't need to call any function of the `AudioEngine` directly. Most stuff is wrapped in `CoreActionController` which should provided a more smooth and accessible interface. 

> Would that even be possible [..]?

I think that's entirely possible but I have no idea how much work it will require (as I can not judge the LV2 part). I would also like to encourage you to give it a shot, report back from time to time, and to feel free to ask questions in case you are stuck. AFAIK there are no plans in the dev team to implement this feature from our side in the near future. Not because we dislike the idea but because our main focus right now is on removing bugs and making existing things better/usable instead of including new features.

**Comment by dingodoppelt** [comment]
Hey there,
thanks for the advice! As I said, I'm not a programmer and this task is way over my head looking at what would have to be done. I might still have a go at this later but not right now. I found drumgizmo to be already very close to what I'm after and it seems more reasonable trying to get drumgizmo to do some of the things I enjoy so much about hydrogen than the other way round.
